# Introduction to Probability

Event is the outcome of a particular random experiment\
Sample space represents the universe of all possible outcomes\
Event space is a subset of sample space where the collection of events we about are\


### The law of large numbers

The probability of the event (P(E)) equals the limit as the number of experiments n goes to infinity of the count of successful experiments S(n) divided by the number of experiments n

<math xmlns="http://www.w3.org/1998/Math/MathML" display="block">
  <mi>P</mi>
  <mo stretchy="false">(</mo>
  <mi>E</mi>
  <mo stretchy="false">)</mo>
  <mo>=</mo>
  <munder>
    <mo data-mjx-texclass="OP" movablelimits="true">lim</mo>
    <mrow data-mjx-texclass="ORD">
      <mi>n</mi>
      <mo stretchy="false">&#x2192;</mo>
      <mi mathvariant="normal">&#x221E;</mi>
    </mrow>
  </munder>
  <mstyle displaystyle="true">
    <mfrac>
      <mrow>
        <mi>S</mi>
        <mrow data-mjx-texclass="ORD">
          <mo stretchy="false">(</mo>
          <mi>n</mi>
          <mo stretchy="false">)</mo>
        </mrow>
      </mrow>
      <mi>n</mi>
    </mfrac>
  </mstyle>
</math>


E = event\
P(E) = Probability of event E\
n = Number of conducted experiments\
S(n) = Count of successful experiments

### Probability Axioms

#### 1. Positivity

A probability is always bigger than or equal to 0, or 0 <= P(E) <= 1

#### 2. Probability of a certain event

If the event of interest is the sample space (S = E), we say that the outcome is a certain event, or P(S) = 1

#### 3. Additivity

The probability of the union of 2 exclusive events is equals to the sum of the probabilities of the individual events happening.\
If the events are exclusive, they never happen at the same time, so the intersection between them is empty. You can simply add the probabilities together.\
If <math xmlns="http://www.w3.org/1998/Math/MathML">
  <mi>A</mi>
  <mo>&#x2229;</mo>
  <mi>B</mi>
  <mo>=</mo>
  <mi mathvariant="normal">&#x2205;</mi>
</math> , then <math xmlns="http://www.w3.org/1998/Math/MathML">
  <mi>P</mi>
  <mo stretchy="false">(</mo>
  <mi>A</mi>
  <mo>&#x222A;</mo>
  <mi>B</mi>
  <mo stretchy="false">)</mo>
  <mo>=</mo>
  <mi>P</mi>
  <mo stretchy="false">(</mo>
  <mi>A</mi>
  <mo stretchy="false">)</mo>
  <mo>+</mo>
  <mi>P</mi>
  <mo stretchy="false">(</mo>
  <mi>B</mi>
  <mo stretchy="false">)</mo>
</math>

##### Example

<math xmlns="http://www.w3.org/1998/Math/MathML">
  <mi>P</mi>
  <mo stretchy="false">(</mo>
  <mi>M</mi>
  <mo>&#x222A;</mo>
  <mi>N</mi>
  <mo stretchy="false">)</mo>
  <mo>=</mo>
  <mi>P</mi>
  <mo stretchy="false">(</mo>
  <mi>M</mi>
  <mo stretchy="false">)</mo>
  <mo>+</mo>
  <mi>P</mi>
  <mo stretchy="false">(</mo>
  <mi>N</mi>
  <mo stretchy="false">)</mo>
  <mo>=</mo>
  <mstyle displaystyle="true">
    <mfrac>
      <mn>1</mn>
      <mn>6</mn>
    </mfrac>
  </mstyle>
  <mo>+</mo>
  <mstyle displaystyle="true">
    <mfrac>
      <mn>3</mn>
      <mn>6</mn>
    </mfrac>
  </mstyle>
  <mo>=</mo>
  <mstyle displaystyle="true">
    <mfrac>
      <mn>4</mn>
      <mn>6</mn>
    </mfrac>
  </mstyle>
  <mo>=</mo>
  <mstyle displaystyle="true">
    <mfrac>
      <mn>2</mn>
      <mn>3</mn>
    </mfrac>
  </mstyle>
</math>

### Addition Law of Probability

The probability that A or B will happen is the sum of the probabilities that A will happen and that B will happen, minus the probability that both A and B will happen.

<math xmlns="http://www.w3.org/1998/Math/MathML">
  <mi>P</mi>
  <mo stretchy="false">(</mo>
  <mi>A</mi>
  <mo>&#x222A;</mo>
  <mi>B</mi>
  <mo stretchy="false">)</mo>
  <mo>=</mo>
  <mi>P</mi>
  <mo stretchy="false">(</mo>
  <mi>A</mi>
  <mo stretchy="false">)</mo>
  <mo>+</mo>
  <mi>P</mi>
  <mo stretchy="false">(</mo>
  <mi>B</mi>
  <mo stretchy="false">)</mo>
  <mo>&#x2212;</mo>
  <mi>P</mi>
  <mo stretchy="false">(</mo>
  <mi>A</mi>
  <mo>&#x2229;</mo>
  <mi>B</mi>
  <mo stretchy="false">)</mo>
</math>

We use this law as exclusive events are not the norm

##### Example

<math xmlns="http://www.w3.org/1998/Math/MathML">
  <mi>P</mi>
  <mo stretchy="false">(</mo>
  <mi>N</mi>
  <mo>&#x222A;</mo>
  <mi>Q</mi>
  <mo stretchy="false">)</mo>
  <mo>=</mo>
  <mi>P</mi>
  <mo stretchy="false">(</mo>
  <mi>N</mi>
  <mo stretchy="false">)</mo>
  <mo>+</mo>
  <mi>P</mi>
  <mo stretchy="false">(</mo>
  <mi>Q</mi>
  <mo stretchy="false">)</mo>
  <mo>&#x2212;</mo>
  <mi>P</mi>
  <mo stretchy="false">(</mo>
  <mi>N</mi>
  <mo>&#x2229;</mo>
  <mi>Q</mi>
  <mo stretchy="false">)</mo>
  <mo>=</mo>
  <mstyle displaystyle="true">
    <mfrac>
      <mn>3</mn>
      <mn>6</mn>
    </mfrac>
  </mstyle>
  <mo>+</mo>
  <mstyle displaystyle="true">
    <mfrac>
      <mn>3</mn>
      <mn>6</mn>
    </mfrac>
  </mstyle>
  <mo>&#x2212;</mo>
  <mstyle displaystyle="true">
    <mfrac>
      <mn>1</mn>
      <mn>6</mn>
    </mfrac>
  </mstyle>
  <mo>=</mo>
  <mstyle displaystyle="true">
    <mfrac>
      <mn>5</mn>
      <mn>6</mn>
    </mfrac>
  </mstyle>
</math>

# LABS

###  Sample space, event space and the law of relative frequency

#### a. Let's throw a dice once

First, create a set `roll_dice` that holds the sample space of rolling a 6-sided dice once.

In [1]:
roll_dice = {1,2,3,4,5,6}

Let's assume that the event space is defined by 'throwing a number higher than 4'\
Successful outcomes are 5 and 6

In [2]:
event = {5,6}

Using the formula P(E) = |E| / |S|(Laplace Formula, related strongly to the law of relative frequency) to calculate probability.

In [3]:
prob_5_6 = len(event)/len(roll_dice)
prob_5_6  # 0.3333333333333333

0.3333333333333333

#### b. Now, let's simulate rolling dice to see how the law of relative frequency works.

$$P(E) = \lim_{n\rightarrow\infty} \dfrac{S{(n)}}{n}$$

As you can see in the formula, the law of relative frequency states that when repeating an experiment $n$ times, where $n$ is very big, and you divide the number of "good" outcomes by the sample space (here we call it event E), you get to the probability of the event E. It should be clear that we get a more accurate number for P(E) when $n$ grows.

First, let's randomly generate values between 1 and 6.

In [4]:
import numpy as np
np.random.randint(1,7) # you will get a random value between 1 and 6. See how it changes when you rerun

3

Now, let's repeat this experiment 10 times, then 1000 times, then 1 million times, then 100 million times. 

In [5]:
np.random.seed(12345) # to ensure reproducibility of results

dice_10 = np.random.randint(1,7,size= 10)
dice_1k = np.random.randint(1,7,size= 1000)
dice_1m = np.random.randint(1,7,size=1000000)
dice_100m = np.random.randint(1,7,size=100000000)

Next, let's count the number of "events". Remember that an event here is defined as throwing a 5 or a 6. Store them in the values below.

In [6]:
event_10 = np.sum((dice_10 == 5) | (dice_10 == 6))
event_1k = np.sum((dice_1k == 5) | (dice_1k == 6))
event_1m = np.sum((dice_1m == 5) | (dice_1m == 6))
event_100m = np.sum((dice_100m == 5) | (dice_100m == 6))

Next, you'll divide the number of events for each  𝑛
  by the respective values for  𝑛
 . What do you see?

In [7]:
prob_10   = event_10 / len(dice_10)
prob_1k   = event_1k / len(dice_1k )
prob_1m   = event_1m / len(dice_1m)
prob_100m = event_100m / len(dice_100m)
prob_10, prob_1k, prob_1m, prob_100m  # 0.5 0.331 0.333657 0.33329752

prob_10, prob_1k, prob_1m, prob_100m  # 0.5 0.331 0.333657 0.33329752

(0.5, 0.331, 0.333657, 0.33329752)

You see that the probability converges to 0.3333333... for higher values of $n$. 

##  The Probability Axioms

You're working at the United Nations, and want to get a better sense of the world population. 

You come across some numbers and find the list of probabilities of being an inhabitant for each of the seven continents (rounded up to 3 digits):

- P(Africa) = 0.161
- P(Antarctica) = 0.000
- P(Asia) = 0.598
- P(Europe) = 0.10
- P(North-America) = 0.079
- P(Australia) = 0.005
- P(South-America) = 0.057

store these values using the variable names below:

In [8]:
P_afr = 0.161
P_ant = 0.000
P_as = 0.598
P_eur = 0.10
P_na = 0.079
P_aus = 0.005
P_sa = 0.057

Now create the sample space set names `continents`. Store the sample space in a numpy array.

In [9]:
continents = np.array([P_afr, P_ant, P_as, P_eur, P_na, P_aus, P_sa])
print(continents)

[0.161 0.    0.598 0.1   0.079 0.005 0.057]


We want to make sure that the three probability axioms are fulfilled because they assure us that $(\Omega,E,P)$ is a **probability space**:

- if we have a sample space $S$ (or $\Omega$)
- if we have an event space $E$ and a probability measure $P$, 
- **and** the three probability axioms introduced in the previous lesson are fulfilled, 

The third axiom is fairly ad hoc, and you will basically have to deduce from the context whether individual events are independent. It is fairly straightforward, however, that people cannot be inhabitants of two continents at the same time, so for now, we will assume that we're good for axiom three.

However, we can use the numpy array `continents` to verify if axiom 1 and 2 are fulfilled. Create a function `check_axioms` with a single input, `sample_space`, that returns the message "We're good!" if both axiom 1 and 2 are fulfilled, and "Not quite!" if that's not the case.

In [10]:
def check_axioms(sample_space):
    # Check Axiom 1: The probability of any event must be between 0 and 1
    for prob in sample_space:
        if prob < 0 or prob > 1:
            return 'Not quite'
    # Check Axiom 2: The probability of the sample space must be 1
    prob_sum = 0
    for prob in sample_space:
        prob_sum += prob
    if abs(prob_sum - 1) > 1e-10:
        return "Not quite!"
    
    # If both axioms hold, return "We're good!"
    return "We're good!"

Now test your newly created function out on the sample space `continents`:

In [11]:
check_axioms(continents)

"We're good!"

You want to make sure your test returns `"Not quite!"` for the following numpy arrays. Go ahead and test away!

In [12]:
test_1 = np.array([0.05, 0.2, 0.3, 1.01])
test_2 = np.array([0.05, 0.5, 0.6, -0.15])
test_3 = np.array([0.043,0.05,.02,0.3,0.2])
check_axioms(test_1)
check_axioms(test_2)
check_axioms(test_3)

'Not quite!'

## Some more practice on the sample and event spaces

In this exercise, we'll look at possible outcomes when throwing a dice twice. For your convenience, we created the NumPy array below.

Next, we'll compute a couple or probabilities associated with doing this.

In [14]:
import numpy as np
sample_dice = np.array([(1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6),
              (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6),
              (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6),
              (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6),
              (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6),
              (6, 1), (6, 2), (6, 3), (6, 4), (6, 5), (6, 6)])
sample_dice.shape # should be equal to (36,2)


(36, 2)

a. What is the probability of throwing a 5 at least once?\
First, use sample_dice to get "True" values for each time a 5 occurs.

In [16]:
set_5 = sample_dice == (5)
#Next, make sure that you get a value True for each pair where at least one 5 was thrown.
true_5 = np.any(set_5, axis = 1)
print(true_5)

[False False False False  True False False False False False  True False
 False False False False  True False False False False False  True False
  True  True  True  True  True  True False False False False  True False]


Applying the sum() function to this result you can get to the total number of items in the event space. Divide this number by the total number in the sample space to obtain the probability of throwing a 5 at least once.

In [17]:
prob_5 = true_5.sum()/len(sample_dice)
print(prob_5)

0.3055555555555556


b. What is the probability of throwing a 5 or 6 at least once?

In [18]:
set_5 =  sample_dice == (5)
set_6 = sample_dice == (6)
set_5_6 = (set_6 + set_5)
set_any_5_6 = set_5_6.any(axis=1)
print(set_any_5_6) 
prob_5_6 = set_any_5_6.sum()/len(sample_dice)
print(prob_5_6)

[False False False False  True  True False False False False  True  True
 False False False False  True  True False False False False  True  True
  True  True  True  True  True  True  True  True  True  True  True  True]
0.5555555555555556


c. What is the probability of the outcome having a sum of exactly 8?

In [19]:
sum_dice = np.sum(sample_dice, axis = 1)
sum_8 = sum(sum_dice == 8)
prob_sum_8 = sum_8/len(sample_dice)
print(prob_sum_8)

0.1388888888888889


#### Now let's try creating your own event space!
A teaching assistant is holding office hours so students can make appointments. She has 6 appointments scheduled today, 3 by male students, and 3 by female students.

Create a NumPy array of all possible orders of three male and three female students in which the teaching assistant can see students by appointment (please note: only consider gender when creating event space). Create this NumPy array in the same way as we did in the "throwing a dice twice" exercise. It will be quite a bit of typing, as your resulting NumPy array will have a shape (20,6)!

In [21]:
sample_mf= np.array([("M","M","M","F","F","F"), ("M","M","F","M","F","F"), ("M","M","F","F","M","F"),
                     ("M","M","F","F","F","M"), ("M","F","M","M","F","F"), ("M","F","M","F","F","M"),
                     ("M","F","M","F","M","F"), ("M","F","F","M","F","M"), ("M","F","F","M","M","F"),
                     ("M","F","F","F","M","M"), ("F","F","F","M","M","M"), ("F","F","M","F","M","M"), 
                     ("F","F","M","M","F","M"), ("F","F","M","M","M","F"), ("F","M","F","F","M","M"),
                     ("F","M","F","M","M","F"), ("F","M","F","M","F","M"), ("F","M","M","F","M","F"),
                     ("F","M","M","F","F","M"), ("F","M","M","M","F","F") ])
print(sample_mf.shape)
sample_length= len(sample_mf)
print(sample_length)

(20, 6)
20


#### 1. Calculate the probability that at least 2 out of the first 3 appointments are with female students
First, select the first 3 appointment slots and check for "F".

In [24]:
first_3_F = sample_mf[:,:3] == "F"
print(first_3_F)
num_F = np.sum(first_3_F, axis=1)
print(num_F)
F_2plus = np.sum(num_F > 1)
print(F_2plus)
prob_F_2plus = F_2plus.sum()/sample_length
print(prob_F_2plus)

[[False False False]
 [False False  True]
 [False False  True]
 [False False  True]
 [False  True False]
 [False  True False]
 [False  True False]
 [False  True  True]
 [False  True  True]
 [False  True  True]
 [ True  True  True]
 [ True  True False]
 [ True  True False]
 [ True  True False]
 [ True False  True]
 [ True False  True]
 [ True False  True]
 [ True False False]
 [ True False False]
 [ True False False]]
[0 1 1 1 1 1 1 2 2 2 3 2 2 2 2 2 2 1 1 1]
10
0.5


#### 2. Calculate the probability that after 4 appointment slots, all the female students have had an appointment

In [25]:
np.sum((sample_mf[:,4:]== ['M','M']).all(axis=1))/sample_length

0.2

You noticed that coming up with the sample space was probably the most time-consuming part of the exercise, and it would really become unfeasible to write this down for say, 10 or, even worse, 20 appointments in a row. You'll learn about methods that make this process easier soon!

## The Addition Law of Probability
At a supermarket, we randomly select customers, and make notes of whether a certain customer owns a Visa card (event A) or an American Express credit card (event B). Some customers own both cards. You can assume that:

P(A) = 0.5
P(B) = 0.4
both A and B = 0.25.
Compute the probability that a selected customer has at least one credit card.

Compute the probability that a selected customer doesn't own any of the mentioned credit cards.

Compute the probability that a customer only owns VISA card.

(You can use Python here, but you don't have to)

In [26]:
# 1) $P(A\cup B) = P(A) + P(B) - P(A\cap B)  = 0.5+0.4-0.25 = 0.65 $

# 2) $1 - P(A\cup B) = 0.35 $

# 3) $P(A)-P(B\cap A) = 0.25$